README
MooseX::Attribute::ValidateWithException

ALPHA QUALITY SOFTWARE.

At present, when an attribute fails validation, Moose internally die()'s
with a string. There is also no way to throw an exception object as part
of the validation message, ( in order to give more context on the
problem ), without also breaking how much of the validation works.

This module is an experiment in providing that feature, which really
should be done in Moose itself, and done better, which is why it has
been given such an obtuse name.

This module makes no promises of forwards compatibility with a future
Moose release, in order to permit Moose to do whatever they want and not
worry about "breaking code" that uses this module. ( So that they can
easily replace this module in incompatible ways )

If your code breaking is unacceptable, do *not* use this module.

Use of this module assumes several things.

1. You are o.k. with your code breaking in a future Moose release.
2. You are o.k. with re-writing any and all code that depends on this
functionality, if a future Moose release is incompatible with this
module.

I'm not saying I won't do my best to provide forwards compatibility, but
it is highly unlikely it will be possible, due to differences in package
naming which may be essential for handling exceptions.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Build.PL
  ./Build
  ./Build test
  ./Build install

COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for MooseX-Attribute-ValidateWithException

v0.4.1 2017-03-12T02:45:26Z 0fbc10c

v0.4.0 2017-03-12T02:32:28Z b0eb239
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot
 - Migrate to EUMM for installer ( may break downstreams, reason for v0.4 )

 [Dependencies::Stats]
 - Dependencies changed since 0.3.1, see misc/*.deps* for details
 - build: -1
 - configure: +2 -1 (recommends: +1)
 - develop: +22 ↑3 -11 (recommends: +1, suggests: +2 -1)
 - runtime: (recommends: +1)
 - test: +3 ↓1 (recommends: +3)

0.3.1 2014-01-08T04:40:40Z
 [00 Minor]
 - Functionally the same as 0.3.0 with -TRIAL removed.
 - Minor docs improvements.

 [Dependencies::Stats]
 - Dependencies changed since 0.3.0, see misc/*.deps* for details
 - develop: +3 ↑1 -4
 - runtime: ↑1

0.3.0 2013-11-28T16:41:47Z
 [00 Major]
 - Refactored to work better with Moose 2.1100
 - Code is now mostly a compatibility shim, because Moose has exceptions default

 [Dependencies::Noteworthy]
 - now requires Moose >2.1100
 - now requires Module::Build 0.4203
 - now requires Test::More 1.001002

 [Dependencies::Stats]
 - Dependencies changed since 0.2.2, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +45 -1 (recommends: -1, suggests: +1 -1)
 - runtime: ↑1
 - test: ↑1 -2

0.2.2 2013-04-10T07:27:10Z
 - Maint release

 [Dependencies::Noteworthy]
 - use Module::Build 0.4004
 - use Test::More 0.98

 [Dependencies::Stats]
 - Dependencies changed since 0.2.1, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +5 (suggests: ↑1)
 - test: ↑1

 [Documentation]
 - Update copyright year.
 - add README.mkdn

 [Internals]
 - fix a warning about a value being undefined by making it lazy

 [Meta]
 - Bug tracker to github issues

 [Packaging]
 - use test_requires in Build.PL

0.2.1 2012-03-17T04:49:40Z
 - Minor maintenance release

 [Dependencies::Stats]
 - Dependencies changed since 0.2.0, see misc/*.deps* for details
 - develop: (recommends: ↑1, suggests: ↑1)
 - runtime: +3
 - test: -1

 [Internals]
 - Version declaration moved out of BEGIN
 - All namespaces declare $AUTHORITY

 [Packaging]
 - Update LICENSE ( Year, Indent, Address )
 - Move author/release tests to xt/
 - set x_authority

0.2.0 2011-07-26T00:18:24Z
 [Bugfix]
 - Fix compatibility with Moose 2.020???. This hopefully fixes rt#60461 (
   https://rt.cpan.org/Public/Bug/Display.html?id=69461 ), thanks ANDK for reporting.

 [Dependencies::Noteworthy]
 - upgrade to Moose 2.0200

 [Dependencies::Stats]
 - Dependencies changed since 0.1.1, see misc/*.deps* for details
 - runtime: ↑1

 [Packaging]
 - Bumping minor version due to dependency changes being not reverse compatible.

0.1.1 2011-06-18T02:24:30Z
 [Bugfix]
 - Now actually propagates exceptions thrown from inlined attributes.

0.1.0 2011-06-18T00:27:59Z
 - First version.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Cause validation failures to throw exception objects.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "MooseX-Attribute-ValidateWithException",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "MooseX::Attribute::ValidateWithException",
  "PREREQ_PM" => {
    "Moose" => "2.1200",
    "Moose::Exporter" => 0,
    "Moose::Role" => 0,
    "Throwable::Error" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Fatal" => 0,
    "Test::More" => 0
  },
  "VERSION" => "v0.4.1",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moose" => "2.1200",
  "Moose::Exporter" => 0,
  "Moose::Role" => 0,
  "Test::Fatal" => 0,
  "Test::More" => 0,
  "Throwable::Error" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/MooseX/Attribute/ValidateWithException.pm
## no critic (Moose::RequireMakeImmutable)
use 5.006;    # warnings
use strict;
use warnings;

package MooseX::Attribute::ValidateWithException;

our $VERSION = 'v0.4.1';

# ABSTRACT: Cause validation failures to throw exception objects.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

require Moose;
use Moose::Exporter;
require MooseX::Attribute::ValidateWithException::AttributeRole;

Moose::Exporter->setup_import_methods(
  class_metaroles => { attribute => ['MooseX::Attribute::ValidateWithException::AttributeRole'], }, );






























































1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MooseX::Attribute::ValidateWithException - Cause validation failures to throw exception objects.

=head1 VERSION

version v0.4.1

=head1 SYNOPSIS

  {
    package Foo;
    use Moose;
    use MooseX::Attribute::ValidateWithException;

    has foo => (
      isa => 'Str',
      is  => 'rw',
      required => 1,
    );
    __PACKAGE__->meta->make_immutable;
    no Moose;
  }

  use Try::Tiny;

  try {
    Foo->new( foo => { this_is => [qw( not what we were wanting )] } );
  } catch {
    say $_->name if blessed( $_ ) && $_->isa('Thingy');
  };

=head1 DESCRIPTION

B<ALPHA QUALITY SOFTWARE>.

At present, when an attribute fails validation, Moose internally die()'s with a
string. There is also no way to throw an exception object as part of the
validation message, ( in order to give more context on the problem ), without
also breaking how much of the validation works.

This module is an experiment in providing that feature, which really should be
done in Moose itself, and done better, which is why it has been given such an
obtuse name.

This module makes no promises of forwards compatibility with a future Moose
release, in order to permit Moose to do whatever they want and not worry about
"breaking code" that uses this module. ( So that they can easily replace this
module in incompatible ways )

B<< If your code breaking is unacceptable, do I<not> use this module >>.

Use of this module assumes several things.

=over 4

=item 1. You are o.k. with your code breaking in a future Moose release.

=item 2. You are o.k. with re-writing any and all code that depends on this
functionality, if a future Moose release is incompatible with this module.

=back

I'm not saying I won't do my best to provide forwards compatibility, but it is
highly unlikely it will be possible, due to differences in package naming
which may be essential for handling exceptions.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/MooseX/Attribute/ValidateWithException/Exception.pm
## no critic (Moose::RequireMakeImmutable)
use 5.006;    # warnings
use strict;
use warnings;

package MooseX::Attribute::ValidateWithException::Exception;

our $VERSION = 'v0.4.1';

# ABSTRACT: An Exception object to represent "Normal" moose validation failures.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moose qw( extends has );

#with 'StackTrace::Auto';

extends 'Throwable::Error';

has 'attribute_name' => (
  isa      => 'Str',
  required => 1,
  is       => 'ro',
);

has 'data' => (
  is       => 'ro',
  required => 1,
);

has 'constraint_message' => (
  is       => 'ro',
  required => 1,
);

# has 'constraint' => (
#  is       => 'ro',
#  required => 1,
#);

has 'constraint_name' => (
  isa      => 'Str',
  is       => 'ro',
  required => 1,
);

has message => (
  isa     => 'Str',
  is      => 'ro',
  lazy    => 1,
  builder => '_generate_message',
);

sub _generate_message {
  my ($self) = shift;
  return sprintf 'Attribute (%s) does not pass the type constraint because: %s', $self->attribute_name, $self->constraint_message;
}

#with 'StackTrace::Auto';

__PACKAGE__->meta->make_immutable( inline_constructor => 0 );
no Moose;
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MooseX::Attribute::ValidateWithException::Exception - An Exception object to represent "Normal" moose validation failures.

=head1 VERSION

version v0.4.1

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/MooseX/Attribute/ValidateWithException/AttributeRole.pm
use 5.006;    # warnings
use strict;
use warnings;

package MooseX::Attribute::ValidateWithException::AttributeRole;

our $VERSION = 'v0.4.1';

# ABSTRACT: Role to apply attribute properties to an attribute.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moose::Role qw( override );

sub __generate_check {
  my ( $self, %params ) = @_;
  if ( $self->type_constraint->can_be_inlined ) {
    ## no critic (ProtectPrivateSubs)
    return sprintf '( %s )', $self->type_constraint->_inline_check( $params{value} );
  }
  else {
    return sprintf '( %s->( %s ) )', $params{tc}, $params{value};
  }
}

override '_inline_check_constraint' => sub {

  my $self = shift;
  my ( $value, $tc, $message, ) = @_;

  my $attribute_name = quotemeta( $self->name );

  return unless $self->has_type_constraint;

  my $tc_obj  = $self->type_constraint;
  my $tc_name = quotemeta( $tc_obj->name );

  ## no critic ( ProhibitImplicitNewlines RequireInterpolationOfMetachars )

  my $check_code       = $self->__generate_check( value => $value, tc => $tc, );
  my $message_variable = '$message';
  my $get_message_code = ( sprintf 'do { local $_ = %s; %s->( %s ) }', $value, $message, $value, );

  return <<"CHECKCODE";
  if( ! $check_code ) {
    my $message_variable = $get_message_code;
    if ( ref $message_variable ) {
      die $message_variable;
    } else {
      require MooseX::Attribute::ValidateWithException::Exception;
      die MooseX::Attribute::ValidateWithException::Exception->new(
        attribute_name     => '$attribute_name',
        data               => $value,
        constraint_message => $message_variable,
        constraint_name    => '$tc_name',
      );
    }
  }
CHECKCODE
};

override 'verify_against_type_constraint' => sub {
  my $self = shift;
  my $val  = shift;

  return 1 if !$self->has_type_constraint;

  my $type_constraint = $self->type_constraint;

  if ( not $type_constraint->check($val) ) {
    my $message = $type_constraint->get_message($val);
    if ( ref $message ) {
      die $message;
    }
    else {
      require MooseX::Attribute::ValidateWithException::Exception;
      die MooseX::Attribute::ValidateWithException::Exception->new(
        attribute_name     => $self->name,
        data               => $val,
        constraint_message => $message,
        constraint         => $type_constraint,
        constraint_name    => $type_constraint->name,
      );
    }
  }
};

no Moose::Role;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MooseX::Attribute::ValidateWithException::AttributeRole - Role to apply attribute properties to an attribute.

=head1 VERSION

version v0.4.1

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.009.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4220"
  },
  "configure_requires" => {
    "Module::Build" => "0.4220",
    "perl" => "5.006"
  },
  "dist_abstract" => "Cause validation failures to throw exception objects.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "MooseX-Attribute-ValidateWithException",
  "dist_version" => "v0.4.0",
  "license" => "perl",
  "module_name" => "MooseX::Attribute::ValidateWithException",
  "recommends" => {
    "Moose" => "2.000"
  },
  "recursive_test_files" => 1,
  "requires" => {
    "Moose" => "2.1200",
    "Moose::Exporter" => 0,
    "Moose::Role" => 0,
    "Throwable::Error" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "1.302073",
    "perl" => "5.006"
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Build" => "0.4220",
  "Test::Fatal" => 0,
  "Test::More" => "1.302073",
  "perl" => "5.006"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);


$build->create_build_script;
